Lecture 3

HidJaba-hub opened the pull request.

HidJaba-hub committed:
Lecture 3

HidJaba-hub committed:
Lecture 3

HidJaba-hub committed:
-fixed whitespace

HidJaba-hub committed:
Merge remote-tracking branch 'origin/lecture-3' into lecture-3

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/entity/ProductHistory.java:
```diff
@@ -0,0 +1,18 @@
+@Setter
+public class ProductHistory {
+
+    Deque<Product> recentlyViewedProducts;
```
access modifier

HidJaba-hub commented on src/main/java/com/es/phoneshop/DAO/CustomProductDao.java:
```diff
@@ -2,8 +2,8 @@
 import com.es.phoneshop.SortOrder;
-import com.es.phoneshop.StringChecker;
 import com.es.phoneshop.model.entity.Product;
+import com.es.phoneshop.utils.StringChecker;
```
123132

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/entity/cart/Cart.java:
```diff
@@ -0,0 +1,21 @@
+@Getter
+public class Cart {
+
+    LinkedHashSet<CartItem> items;
```
access modifier

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public Cart getCart(HttpServletRequest request) {
+        HttpSession session = (HttpSession) SyncObjectPool.getSyncObject(request.hashCode(), request.getSession());
```
session has its own id, no need to check the request.hashCode()

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/utils/SyncObjectPool.java:
```diff
@@ -0,0 +1,37 @@
+                        lastObjectKey = idToCheck;
+                    }
+                } finally {
+                    if (syncObjects.size() < MAX_POOL_SIZE) lock.unlock();
```
formatting, braces

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:
```diff
@@ -10,26 +14,64 @@
+                cartService.addProductToCart(cart, productId, Integer.valueOf(quantity));
+            } catch (OutOfStockException e) {
+                String errorString;
+                if (e.getStockAvailable() == 0) errorString = "Wrong amount of products";
```
formatting and braces

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ViewedProductFilter.java:
```diff
@@ -0,0 +1,47 @@
+
+import java.io.IOException;
+
+public class ViewedProductFilter implements Filter {
```
why filter?

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/service/CustomCartServiceTest.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Test(expected = OutOfStockException.class)
+    public void givenNegativeQuantity_whenAddProduct_thenGetException() throws OutOfStockException {
```
in case of negative stock we should throw smth like IllegalArgumentException, but not OutOfStock for cure

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/service/CustomProductHistoryTest.java:
```diff
@@ -0,0 +1,84 @@
+        when(request.getSession().getAttribute(any())).thenReturn(productHistory);
+
+        ProductHistory testProductHistory = productHistoryService.getProductHistory(request);
+        ;
```
;

HidJaba-hub committed:
-fixed SyncObjectPool

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public void addProductToCart(Cart cart, Long productId, int quantity) throws OutOfStockException {
+        cart = (Cart) SyncObjectPool.getSyncObject(cart.hashCode(), cart);
```
Cart should have its own id, as it's done for the Product

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public Cart getCart(HttpServletRequest request) {
+        Object syncObject = SyncObjectPool.getSyncObject(request.hashCode());
```
getSyncObject(sessionId)

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public void addProductToCart(Cart cart, Long productId, int quantity) throws OutOfStockException {
+        Object syncObject = SyncObjectPool.getSyncObject(cart.hashCode());
```
getSyncObject(cartId)
Cart should have it's own id as it's done for Product

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/productHistory/CustomProductHistoryService.java:
```diff
@@ -0,0 +1,64 @@
+    @Override
+    public ProductHistory getProductHistory(HttpServletRequest request) {
+        HttpSession session = request.getSession();
+        Object syncObject = SyncObjectPool.getSyncObject(session.hashCode());
```
session id

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/productHistory/CustomProductHistoryService.java:
```diff
@@ -0,0 +1,64 @@
+        synchronized (syncObject) {
+            ProductHistory productHistory = (ProductHistory) session.getAttribute(VIEWED_PRODUCTS_SESSION_ATTRIBUTE);
+            if (productHistory == null) {
+                request.getSession().setAttribute(VIEWED_PRODUCTS_SESSION_ATTRIBUTE, productHistory = new ProductHistory());
```
setAttribute(VIEWED_PRODUCTS_SESSION_ATTRIBUTE, productHistory = new ProductHistory()) ?

esSoftPalinaTsi replied:
and why getProductHistory sets something

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/productHistory/CustomProductHistoryService.java:
```diff
@@ -0,0 +1,64 @@
+        Object syncObject = SyncObjectPool.getSyncObject(session.hashCode());
+        synchronized (syncObject) {
+            ProductHistory productHistory = (ProductHistory) session.getAttribute(VIEWED_PRODUCTS_SESSION_ATTRIBUTE);
+            if (productHistory == null) {
```
how about session listener?

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/productHistory/CustomProductHistoryService.java:
```diff
@@ -0,0 +1,64 @@
+
+    @Override
+    public void addViewedProduct(ProductHistory viewedProducts, Product product) {
+        Object syncObject = SyncObjectPool.getSyncObject(viewedProducts.hashCode());
```
can be synced with session id param

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ViewedProductFilter.java:
```diff
@@ -0,0 +1,47 @@
+        if (servletRequest != null && (request).getMethod().equalsIgnoreCase("GET")
+                && !StringUtils.isEmpty(request.getPathInfo())) {
+            ProductHistory productHistory = productHistoryService.getProductHistory(request);
+            request.setAttribute("viewedProducts", productHistoryService.getProductHistory(request));
```
why request?

esSoftPalinaTsi commented on src/main/webapp/WEB-INF/tags/viewedProducts.tag:
```diff
@@ -0,0 +1,24 @@
+<style><%@ include file="/styles/main.css"%></style>
+
+<c:if test="${viewedProducts != null}">
+<jsp:useBean id="viewedProducts" type="com.es.phoneshop.model.entity.ProductHistory" scope="session"/>
```
in filter you saved data to the request
then in service to session
and here we have scope="session"
something goes wrong within the implementation of recently viewed, please refactor

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,80 @@
+        Object syncObject = SyncObjectPool.getSyncObject(request.hashCode());
+        synchronized (syncObject) {
+            Cart cart = (Cart) request.getSession().getAttribute(CART_SESSION_ATTRIBUTE);
+            if (cart == null) {
```
how about listener?

HidJaba-hub committed:
Fixed issues:
-removed filter
-reformatted
-added listeners

HidJaba-hub committed:
Fixed issues:
-added init params
-expanded tests
-reformatted
-removed filter
-added listeners

HidJaba-hub committed:
Merge remote-tracking branch 'origin/lecture-3' into lecture-3

# Conflicts:
#	src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java
#	src/main/java/com/es/phoneshop/service/productHistory/CustomProductHistoryService.java
#	src/main/java/com/es/phoneshop/web/listener/CartSessionListener.java
#	src/main/java/com/es/phoneshop/web/listener/ProductHistorySessionListener.java
#	src/main/webapp/styles/main.css
#	src/test/java/com/es/phoneshop/web/listener/CartSessionListenerTest.java
#	src/test/java/com/es/phoneshop/web/listener/ProductHistorySessionListenerTest.java

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:
```diff
@@ -36,42 +41,63 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+        String errorString;
+        Cart cart = cartService.getCart(request);
+        try {
+            cartService.addProductToCart(cart, productId, Integer.valueOf(quantity));
```
int quantity

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:
```diff
@@ -36,42 +41,63 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
                 request.setAttribute("product", productService.getProductById(productId));
                 request.setAttribute("cart", cartService.getCart(request));
-            } // compressing field of view
+                makeRecentlyViewedProductList(request, productId);
```
fillRecentlyViewedProductList

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:
```diff
@@ -36,42 +41,63 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
-                return;
-            }
-            response.sendRedirect(request.getContextPath() + "/products/" + productId + "?message=Product added to cart");
+            int quantity = checkQuantity(request, response, quantityStr, productId);
```
parseQuantity
quantityStr can be retrieved inside private parseQuantity

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:
```diff
@@ -36,42 +41,63 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+
+    private void makeRecentlyViewedProductList(HttpServletRequest request, long productId) {
+        ProductHistory productHistory = productHistoryService.getProductHistory(request);
+        request.setAttribute("viewedProducts", productHistoryService.getProductHistory(request));
```
you used session listener for initialization and in jsp this:
<jsp:useBean id="viewedProducts" type="com.es.phoneshop.model.entity.ProductHistory" **scope="session"**/>

then why do we need save "viewedProducts" to request? They should be saved and updated in session

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/productHistory/CustomProductHistoryService.java:
```diff
@@ -24,19 +24,15 @@ public static CustomProductHistoryService getInstance() {
     }
 
     @Override
     public void addViewedProduct(ProductHistory viewedProducts, Product product) {
```
are these updates saved within the session?

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/listener/ProductHistorySessionListener.java:
```diff
@@ -0,0 +1,34 @@
+public class ProductHistorySessionListener implements HttpSessionListener {
+
+    private static final String VIEWED_PRODUCTS_SESSION_ATTRIBUTE = "viewedProducts";
+    private HttpSession session;
```
no need to store as a local var

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/listener/CartSessionListener.java:
```diff
@@ -0,0 +1,35 @@
+public class CartSessionListener implements HttpSessionListener {
+
+    private static final String CART_SESSION_ATTRIBUTE = DefaultCartService.class.getName() + "cart";
+    private HttpSession session;
```
no need to store as a local var

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/productHistory/CustomProductHistoryService.java:
```diff
@@ -24,19 +24,15 @@ public static CustomProductHistoryService getInstance() {
     @Override
     public ProductHistory getProductHistory(HttpServletRequest request) {
```
let's improve namings a bit
productHistory looks confusing, maybe smth like getViewedProductsHistory, to epmhasize that it's a list of visited products

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/entity/ProductHistory.java:
```diff
@@ -0,0 +1,21 @@
+
+@Getter
+@Setter
+public class ProductHistory {
```
let's improve the naming here as well to make it clear that it's about viewed products

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/entity/cart/CartItem.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof CartItem cartItem)) return false;
```
if (this == o) {
return true;
}

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:
```diff
@@ -10,26 +17,87 @@
+            {
+                long productId = Long.valueOf(request.getPathInfo().substring(1));
+                request.setAttribute("product", productService.getProductById(productId));
+                request.setAttribute("cart", cartService.getCart(request));
```
same as for viewed products: you saved cart in session

HidJaba-hub committed:
Fixed issues:
-reformatted
-fixed class namings
-cart and RecentlyViewedProducts now added to session
-fixed whitespaces

HidJaba-hub committed:
Fixed issues:
-reformatted
-fixed class namings
-cart and RecentlyViewedProducts now added to session

HidJaba-hub committed:
-fixed whitespaces

HidJaba-hub committed:
Merge remote-tracking branch 'origin/lecture-3' into lecture-3

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:
```diff
@@ -10,26 +17,83 @@
+            {
+                long productId = Long.valueOf(request.getPathInfo().substring(1));
+                request.setAttribute("product", productService.getProductById(productId));
+                request.getSession().setAttribute("cart", cartService.getCart(request));
```
we already saved cart to session in CartSessionListener

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/service/CustomCartServiceTest.java:
```diff
@@ -0,0 +1,110 @@
+
+        Cart testCart = defaultCartService.getCart(request);
+
+        assertEquals(testCart, cart);
```
in assertions first object is expected, the second is actual

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/service/CustomRecentlyViewedProductsTest.java:
```diff
@@ -0,0 +1,83 @@
+
+        RecentlyViewedProducts testRecentlyViewedProducts = productHistoryService.getRecentlyViewedProducts(request);
+
+        assertEquals(testRecentlyViewedProducts, recentlyViewedProducts);
```
order of objects

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/web/listener/CartSessionListenerTest.java:
```diff
@@ -0,0 +1,61 @@
+
+        cartSessionListener.sessionCreated(sessionEvent);
+
+        verify(session, atLeast(0)).setAttribute(any(), any());
```
atLeast(0) looks weird, never() could be a valid replacement

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/web/listener/RecentlyViewedProductsSessionListenerTest.java:
```diff
@@ -0,0 +1,57 @@
+
+        productHistorySessionListener.sessionCreated(sessionEvent);
+
+        verify(session, atLeast(0)).setAttribute(any(), any());
```
atLeast(0) looks weird, never() could be a valid replacement

esSoftPalinaTsi commented on src/main/webapp/WEB-INF/pages/productDetails.jsp:
```diff
@@ -2,28 +2,65 @@
+                ${param.message}
+        </div>
+    </c:if>
+    <c:if test="${not empty param.error}">
```
in case of error customer doesn't see the invalid quantity
can we add this ability? To message or save in input field

HidJaba-hub committed:
-fixed assertions
-added wrong output amount of products
